mpc/mpc-aio/src/protocol/garble/exec/dual.rs

```diff
@@ -12,287 +12,569 @@ use crate::protocol::{
+use futures::{future::ready, SinkExt, StreamExt};
+use mpc_circuits::{Circuit, Input, InputValue, OutputValue, WireGroup};
 use mpc_core::garble::{
-    exec::dual as core, gc_state, ActiveInputLabels, Delta, FullInputLabels, GarbledCircuit,
+    exec::dual as core, gc_state, ActiveInputLabels, ActiveInputLabelsSet, FullInputLabels,
+    FullInputLabelsSet, GarbledCircuit,
 };
 use utils_aio::expect_msg_or_err;
 
-pub struct DualExLeader<B, S, R>
+mod state {
+    use super::*;
+
+    mod sealed {
+        pub trait Sealed {}
+
+        impl Sealed for super::Initialized {}
+        impl Sealed for super::LabelSetup {}
+    }
+
+    pub trait State: sealed::Sealed {}
+
+    pub struct Initialized;
+
+    pub struct LabelSetup {
+        pub(crate) gen_labels: FullInputLabelsSet,
+        pub(crate) ev_labels: ActiveInputLabelsSet,
+    }
+
+    impl State for Initialized {}
+    impl State for LabelSetup {}
+}
+
+use state::*;
+
+pub struct DualExLeader<S, B, LS, LR>
 where
+    S: State,
     B: Generator + Evaluator,
-    S: ObliviousSend<FullInputLabels>,
-    R: ObliviousReceive<InputValue, ActiveInputLabels>,
+    LS: ObliviousSend<FullInputLabels>,
+    LR: ObliviousReceive<InputValue, ActiveInputLabels>,
 {
+    state: S,
+    circ: Arc<Circuit>,
     channel: GarbleChannel,
     backend: B,
-    label_sender: S,
-    label_receiver: R,
+    label_sender: Option<LS>,
+    label_receiver: Option<LR>,
 }
 
-impl<B, S, R> DualExLeader<B, S, R>
+impl<B, LS, LR> DualExLeader<Initialized, B, LS, LR>
 where
     B: Generator + Evaluator + Send,
-    S: ObliviousSend<FullInputLabels> + Send,
-    R: ObliviousReceive<InputValue, ActiveInputLabels> + Send,
+    LS: ObliviousSend<FullInputLabels> + Send,
+    LR: ObliviousReceive<InputValue, ActiveInputLabels> + Send,
 {
-    pub fn new(channel: GarbleChannel, backend: B, label_sender: S, label_receiver: R) -> Self {
-        Self {
+    /// Create a new DualExLeader
+    pub fn new(
+        circ: Arc<Circuit>,
+        channel: GarbleChannel,
+        backend: B,
+        label_sender: Option<LS>,
+        label_receiver: Option<LR>,
+    ) -> DualExLeader<Initialized, B, LS, LR> {
+        DualExLeader {
+            state: Initialized,
+            circ,
             channel,
             backend,
             label_sender,
             label_receiver,
         }
     }
+
+    /// Exchange input labels
+    ///
+    /// * `gen_labels` - Labels to garble the leader's circuit
+    /// * `gen_inputs` - Inputs to be sent directly to the follower
+    /// * `ot_send_inputs` - Inputs to be sent via OT
+    /// * `ot_receive_inputs` - Inputs to be received via OT
+    /// * `cached_labels` - Cached input labels for the follower's circuit
+    pub async fn setup_inputs(
+        mut self,
+        gen_labels: FullInputLabelsSet,
+        gen_inputs: Vec<InputValue>,
+        ot_send_inputs: Vec<Input>,
+        ot_receive_inputs: Vec<InputValue>,
+        cached_labels: Vec<ActiveInputLabels>,
```

Comment:
> This will cause an error when creating the ActiveInputLabelsSet but I can add a pre-check.

oh, I see now, it will check for dups later. That check should be enough.

> Why do you say this? The Evaluator caches the Generator's input labels as well

nvm, I see that it is better to put all labels into the cache - both received directly and via OT.